### PR TITLE
Nvidia externalplatform

### PR DESCRIPTION
**Pull request by aleasto**
Add nvidia EGL external platforms x11 and xcb.
Group all nvidia EGL external platform libraries into one part.

### CODE REVIEW

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -145,23 +145,47 @@ parts:
...
       - usr/lib
       - usr/share/doc/*/copyright
-      - usr/share/egl/egl_external_platform.d
 
-  gbm:
-    # GBM EGL external platform
-    plugin: nil
+  nvidia-egl-ext-deps:
+    # eglexternalplatform-dev >= 1.2
+    source: https://github.com/NVIDIA/eglexternalplatform/archive/refs/tags/1.2.tar.gz
```
Comment (Saviq) [inline snap/snapcraft.yaml]: This more git-y, I suppose?

```suggestion
    source: https://github.com/NVIDIA/eglexternalplatform.git
    source-tag: 1.2
    source-depth: 1
```

Will also yield better to SBOM generation with time.

File: snap/snapcraft.yaml [hunk 2]
```diff
@@ -145,23 +145,47 @@ parts:
...
+    plugin: meson
+    build-packages:
+      - meson
+    override-prime: '' # only build headers
+
+  nvidia-egl-ext:
+    # EGL external platforms
+    after: [nvidia-egl-ext-deps]
+    # change to pull egl-x11 from the archive once packaged
+    source: https://github.com/NVIDIA/egl-x11/archive/8aac36c712561ebfecc82af3db15c50cd0d573fb.tar.gz
```
Comment (Saviq) [inline snap/snapcraft.yaml]: Same:

```suggestion
    source: https://github.com/NVIDIA/egl-x11.git
    source-commit: 8aac36c712561ebfecc82af3db15c50cd0d573fb
    source-depth: 1
```

**Review by Saviq — CHANGES_REQUESTED** [review]
Just nitpicks for now.

Thanks!

File: snap/snapcraft.yaml (hunk 1 above)
Comment (aleasto) [reply snap/snapcraft.yaml]: Done

File: snap/snapcraft.yaml (hunk 2 above)
Comment (aleasto) [reply snap/snapcraft.yaml]: Done

File: snap/snapcraft.yaml [hunk 3]
```diff
@@ -145,23 +145,51 @@ parts:
...
     prime:
       - usr/lib
       - usr/share/doc/*/copyright
-      - usr/share/egl/egl_external_platform.d
 
-  gbm:
-    # GBM EGL external platform
-    plugin: nil
+  nvidia-egl-ext-deps:
+    # eglexternalplatform-dev >= 1.2
```
Comment (Saviq) [inline snap/snapcraft.yaml]: I think this comment should go, as it's actually `== 1.2` below, or you'd have to explain more that this _has_ to be `>= 1.2`.

File: snap/snapcraft.yaml [hunk 4]
```diff
@@ -145,23 +145,51 @@ parts:
...
+    source-depth: 1
+    plugin: meson
+    build-packages:
+      - meson
+    override-prime: '' # only build headers
+
+  nvidia-egl-ext:
+    # EGL external platforms
+    after: [nvidia-egl-ext-deps]
+    # change to pull egl-x11 from the archive once packaged
```
Comment (Saviq) [inline snap/snapcraft.yaml]:
```suggestion
    # TODO: change to pull egl-x11 from the archive once packaged
```

**Review by Saviq — CHANGES_REQUESTED** [review]
Testing here: https://launchpad.net/~saviq/+snap/mesa-2404-test

File: snap/snapcraft.yaml (hunk 3 above)
Comment (aleasto) [reply snap/snapcraft.yaml]: Removed

File: snap/snapcraft.yaml (hunk 4 above)
Comment (aleasto) [reply snap/snapcraft.yaml]: Done

**Review by Saviq — APPROVED** [review]
OK, this looks fine!

Do you have a story for testing this out in real life?

**Comment by aleasto** [comment]
> OK, this looks fine!
> 
> Do you have a story for testing this out in real life?

You can test it with the `graphics-test-tools` snap from channel `24/stable`:
```
$ graphics-test-tools.eglinfo -B -p x11
```

If you're looking for real applications, any Qt app can use `EGL_KHR_platform_x11` by setting the `QT_XCB_GL_INTEGRATION=xcb_egl` environment variable. If the desktop is Wayland, you should also make the app use XWayland instead with `QT_QPA_PLATFORM=xcb`.

I know of one _real_ program which does that by default: `xwaylandvideobridge`. But I don't think anybody has packaged it as a snap.

Without this change you should see Zink (if vulkan is available) or llvmpipe be used instead.

**Pull request merged**